gridstatus/isone_api/isone_api.py

```diff
@@ -0,0 +1,439 @@
+            )
+
+        self.sleep_seconds = sleep_seconds
+        self.initial_delay = min(max(0.1, sleep_seconds), 60.0)
+        self.max_retries = min(max(0, max_retries), 10)
+
+    def _handle_end_date(
+        self,
+        date: pd.Timestamp,
+        end: pd.Timestamp | None,
+        days_to_add_if_no_end: int,
+    ) -> pd.Timestamp:
+        """
+        Handles a provided end date by either
+
+        1. Using the provided end date converted to the default timezone
+        2. Adding the number of days to the date and converting to the default timezone
+        """
+        if end:
+            end = utils._handle_date(end, tz=self.default_timezone)
+        else:
+            # Have to convert to UTC to do addition, then convert back to local time
+            # to avoid DST issues
+            end = (
+                (date.tz_convert("UTC") + pd.DateOffset(days=days_to_add_if_no_end))
+                .normalize()
+                .tz_localize(None)
+                .tz_localize(self.default_timezone)
+            )
+
+        return end
+
+    def make_api_call(
+        self,
+        url: str,
+        api_params: dict = None,
+        parse_json: bool = True,
+        verbose: bool = False,
+    ):
+        log.debug(f"Requesting url: {url} with params: {api_params}")
+        retries = 0
+        delay = self.initial_delay
+        headers = {"Accept": "application/json"}
+        while retries <= self.max_retries:
+            response = requests.get(
+                url,
+                params=api_params,
+                auth=(self.username, self.password),
+                headers=headers,
+            )
+
+            retries += 1
+            if response.status_code == 200:
+                break
+            elif response.status_code == 429 and retries <= self.max_retries:
+                log.warn(
+                    f"Warn: Rate-limited: waiting {delay} seconds before retry {retries}/{self.max_retries} "
+                    f"requesting url: {url} with params: {api_params}",
+                    verbose,
+                )
+                time.sleep(delay)
+                delay *= 2
+            else:
+                if retries > self.max_retries:
+                    error_message = (
+                        f"Error: Rate-limited still after {self.max_retries} retries. "
+                        f"Failed to get data from {url} with params: {api_params}"
+                    )
+                else:
+                    error_message = (
+                        f"Error: Failed to get data from {url} with params:"
+                        f" {api_params}"
+                    )
+                log.error(error_message)
+                response.raise_for_status()
+
+        if parse_json:
+            return response.json()
+        else:
+            return response.content
+
+    def get_locations(self) -> pd.DataFrame:
+        """
+        Get a list of all locations.
```

Comment:
Is this all locations or is it only "core" locations? 

> Get list of core locations (Hub, Load Zones, External Nodes)

https://webservices.iso-ne.com/docs/v1.1/rest.locations.html

Reply:
Ah yeah this is a typo, gets a list of core locations. the locations_all endpoint I believe gets "everything", meaning nodal locations since it returns about 1300 locations. 

Updated.